File: app/src/main/java/igrek/todotree/commands/ClipboardCommand.java
package igrek.todotree.commands;


import java.util.Set;
import java.util.TreeSet;

import javax.inject.Inject;

import igrek.todotree.dagger.DaggerIOC;
import igrek.todotree.domain.treeitem.AbstractTreeItem;
import igrek.todotree.domain.treeitem.LinkTreeItem;
import igrek.todotree.domain.treeitem.TextTreeItem;
import igrek.todotree.service.clipboard.SystemClipboardManager;
import igrek.todotree.service.clipboard.TreeClipboardManager;
import igrek.todotree.service.resources.UserInfoService;
import igrek.todotree.service.tree.TreeManager;
import igrek.todotree.service.tree.TreeScrollCache;
import igrek.todotree.service.tree.TreeSelectionManager;
import igrek.todotree.ui.GUI;

public class ClipboardCommand {
	
	@Inject
	TreeManager treeManager;
	
	@Inject
	SystemClipboardManager systemClipboardManager;
	
	@Inject
	TreeClipboardManager treeClipboardManager;
	
	@Inject
	UserInfoService userInfo;
	
	@Inject
	GUI gui;
	
	@Inject
	TreeSelectionManager selectionManager;
	
	@Inject
	TreeScrollCache scrollCache;
	
	public ClipboardCommand() {
		DaggerIOC.getFactoryComponent().inject(this);
	}
	
	public void copyItems(Set<Integer> itemPosistions, boolean info) {
		if (!itemPosistions.isEmpty()) {
			treeClipboardManager.clearClipboard();
			AbstractTreeItem currentItem = treeManager.getCurrentItem();
			treeClipboardManager.setCopiedFrom(currentItem);
			for (Integer selectedItemId : itemPosistions) {
				AbstractTreeItem selectedItem = currentItem.getChild(selectedItemId);
				treeClipboardManager.addToClipboard(selectedItem);
			}
			//if one item selected - copying also to system clipboard
			if (treeClipboardManager.getClipboardSize() == 1) {
				AbstractTreeItem item = treeClipboardManager.getClipboard().get(0);
				systemClipboardManager.copyToSystemClipboard(item.getDisplayName());
				if (info)
					userInfo.showInfo("Item copied: " + item.getDisplayName());
			} else {
				if (info)
					userInfo.showInfo("Items copied: " + treeClipboardManager.getClipboardSize());
			}
			// deselect items
			if (selectionManager.isAnythingSelected()) {
				new ItemSelectionCommand().deselectAll();
			}
		} else {
			if (info)
				userInfo.showInfo("No items to copy.");
		}
	}
	
	public void copySelectedItems() {
		if (selectionManager.isAnythingSelected()) {
			copyItems(selectionManager.getSelectedItems(), true);
		} else {
			userInfo.showInfo("No selected items");
		}
	}
	
	public void cutItems(TreeSet<Integer> itemPosistions) {
		if (!itemPosistions.isEmpty()) {
			copyItems(itemPosistions, false);
			userInfo.showInfo("Items cut: " + itemPosistions.size());
			new ItemTrashCommand().removeItems(itemPosistions, false);
		}
	}
	
	public void cutSelectedItems() {
		if (selectionManager.isAnythingSelected()) {
			cutItems(selectionManager.getSelectedItems());
		} else {
			userInfo.showInfo("No selected items");
		}
	}
	
	public void pasteItems(int position) {
		scrollCache.storeScrollPosition(treeManager.getCurrentItem(), gui.getCurrentScrollPos());
		if (treeClipboardManager.isClipboardEmpty()) {
			String systemClipboard = systemClipboardManager.getSystemClipboard();
			if (systemClipboard != null) {
				//wklejanie 1 elementu z systemowego schowka
				treeManager.addToCurrent(position, new TextTreeItem(systemClipboard));
				userInfo.showInfo("Item pasted: " + systemClipboard);
				new GUICommand().updateItemsList();
				gui.scrollToPosition(scrollCache.restoreScrollPosition(treeManager.getCurrentItem()));
			} else {
				userInfo.showInfo("Clipboard is empty.");
			}
		} else {
			for (AbstractTreeItem clipboardItem : treeClipboardManager.getClipboard()) {
				clipboardItem.setParent(treeManager.getCurrentItem());
				treeManager.addToCurrent(position, clipboardItem);
				position++; // next item pasted below
			}
			userInfo.showInfo("Items pasted: " + treeClipboardManager.getClipboardSize());
			treeClipboardManager.recopyClipboard();
			new GUICommand().updateItemsList();
			gui.scrollToPosition(scrollCache.restoreScrollPosition(treeManager.getCurrentItem()));
		}
	}
	
	public void pasteItemsAsLink(int position) {
		scrollCache.storeScrollPosition(treeManager.getCurrentItem(), gui.getCurrentScrollPos());
		if (treeClipboardManager.isClipboardEmpty()) {
			userInfo.showInfo("Clipboard is empty.");
		} else {
			for (AbstractTreeItem clipboardItem : treeClipboardManager.getClipboard()) {
				treeManager.addToCurrent(position, buildLinkItem(clipboardItem));
				position++; // next item pasted below
			}
			userInfo.showInfo("Items pasted as links: " + treeClipboardManager.getClipboardSize());
			new GUICommand().updateItemsList();
			gui.scrollToPosition(scrollCache.restoreScrollPosition(treeManager.getCurrentItem()));
		}
	}
	
	private AbstractTreeItem buildLinkItem(AbstractTreeItem clipboardItem) {
		if (clipboardItem instanceof LinkTreeItem) { // making link to link
			LinkTreeItem linkItem = (LinkTreeItem) clipboardItem;
			return linkItem.clone();
		}
		LinkTreeItem link = new LinkTreeItem(treeManager.getCurrentItem(), null, null);
		link.setTarget(treeClipboardManager.getCopiedFrom(), clipboardItem.getDisplayName());
		return link;
	}
	
}

File: app/src/main/java/igrek/todotree/intent/ClipboardCommand.kt
package igrek.todotree.intent

import igrek.todotree.domain.treeitem.AbstractTreeItem
import igrek.todotree.domain.treeitem.LinkTreeItem
import igrek.todotree.domain.treeitem.TextTreeItem
import igrek.todotree.info.UiInfoService
import igrek.todotree.info.logger.Logger
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import igrek.todotree.service.clipboard.SystemClipboardManager
import igrek.todotree.service.clipboard.TreeClipboardManager
import igrek.todotree.service.tree.TreeManager
import igrek.todotree.service.tree.TreeSelectionManager
import igrek.todotree.ui.treelist.TreeListLayout
import java.util.TreeSet

class ClipboardCommand(
    treeManager: LazyInject<TreeManager> = appFactory.treeManager,
    systemClipboardManager: LazyInject<SystemClipboardManager> = appFactory.systemClipboardManager,
    treeClipboardManager: LazyInject<TreeClipboardManager> = appFactory.treeClipboardManager,
    uiInfoService: LazyInject<UiInfoService> = appFactory.uiInfoService,
    treeSelectionManager: LazyInject<TreeSelectionManager> = appFactory.treeSelectionManager,
) {
    private val treeManager by LazyExtractor(treeManager)
    private val systemClipboardManager by LazyExtractor(systemClipboardManager)
    private val treeClipboardManager by LazyExtractor(treeClipboardManager)
    private val uiInfoService by LazyExtractor(uiInfoService)
    private val treeSelectionManager by LazyExtractor(treeSelectionManager)
    private val treeListLayout: TreeListLayout by LazyExtractor(appFactory.treeListLayout)

    private val logger: Logger = LoggerFactory.logger

    fun copySelectedItems() {
        if (!treeSelectionManager.isAnythingSelected) {
            return uiInfoService.showInfo("No selected items")
        }
        copyItems(treeSelectionManager.selectedItems!!, true)
    }

    fun copyItems(itemPosistions: Set<Int>, info: Boolean, cut: Boolean = false) {
        if (itemPosistions.isEmpty()) {
            if (info)
                uiInfoService.showInfo("No items to copy")
            return
        }
        treeClipboardManager.clearClipboard()
        val currentItem = treeManager.currentItem
        treeClipboardManager.copiedFrom = currentItem
        treeClipboardManager.markForCut = cut
        for (selectedItemId in itemPosistions) {
            currentItem?.getChild(selectedItemId)?.let { selectedItem ->
                treeClipboardManager.addToClipboard(selectedItem)
            }
        }

        if (!cut) {
            val copiedText = treeClipboardManager.clipboard.joinToString("\n") {
                it.displayName
            }
            systemClipboardManager.copyToSystemClipboard(copiedText)
        }

        if (info) {
            if (treeClipboardManager.clipboardSize == 1) {
                val item = treeClipboardManager.clipboard.first()
                uiInfoService.showInfo("Item copied: ${item.displayName}")
            } else {
                uiInfoService.showInfo("Items copied: ${treeClipboardManager.clipboardSize}")
            }
        }

        if (treeSelectionManager.isAnythingSelected) {
            ItemSelectionCommand().deselectAll()
        }
    }

    fun cutSelectedItems() {
        if (!treeSelectionManager.isAnythingSelected) {
            return uiInfoService.showInfo("No selected items")
        }
        cutItems(treeSelectionManager.selectedItems!!)
    }

    fun cutItems(itemPosistions: TreeSet<Int>) {
        if (itemPosistions.isEmpty()) return
        copyItems(itemPosistions, false, cut=true)
        uiInfoService.showInfo("Marked for cut: " + itemPosistions.size)
    }

    fun pasteItems(aPosition: Int) {
        var position = aPosition
        if (treeClipboardManager.isClipboardEmpty) {
            // recover by taking text from system clipboard
            val systemClipboard = systemClipboardManager.systemClipboard ?: return run {
                uiInfoService.showInfo("Clipboard is empty")
            }
            treeManager.addToCurrent(position, TextTreeItem(systemClipboard))
            uiInfoService.showInfo("Pasted from system clipboard: $systemClipboard")
            treeListLayout.updateItemsList()
            return
        }
        when (treeClipboardManager.markForCut) {
            false -> {
                for (clipboardItem in treeClipboardManager.clipboard) {
                    val newItem = clipboardItem.clone()
                    newItem.setParent(treeManager.currentItem)
                    treeManager.addToCurrent(position, newItem)
                    position++ // paste next items below
                }
                uiInfoService.showInfo("Pasted items: ${treeClipboardManager.clipboardSize}")
                treeClipboardManager.recopyClipboard()
                treeListLayout.updateItemsList()
            }
            true -> {
                for (clipboardItem in treeClipboardManager.clipboard) {
                    val newItem = clipboardItem.clone()
                    val oldParent = clipboardItem.getParent()
                    if (oldParent == null) {
                        logger.warn("item doesn't have a parent")
                        continue
                    }
                    treeManager.removeFromParent(clipboardItem, oldParent)
                    newItem.setParent(treeManager.currentItem)
                    treeManager.addToCurrent(position, newItem)
                    position++ // paste next items below
                }
                uiInfoService.showInfo("Moved items: ${treeClipboardManager.clipboardSize}")
                treeClipboardManager.markForCut = false
                treeClipboardManager.clearClipboard()
                treeListLayout.updateItemsList()
            }
        }
    }

    fun pasteItemsAsLink(aPosition: Int) {
        var position = aPosition
        if (treeClipboardManager.isClipboardEmpty) {
            return uiInfoService.showInfo("Clipboard is empty.")
        }
        for (clipboardItem in treeClipboardManager.clipboard) {
            treeManager.addToCurrent(position, buildLinkItem(clipboardItem))
            position++ // next item pasted below
        }
        treeClipboardManager.markForCut = false
        uiInfoService.showInfo("Items pasted as links: " + treeClipboardManager.clipboardSize)
        treeListLayout.updateItemsList()
    }

    private fun buildLinkItem(clipboardItem: AbstractTreeItem): AbstractTreeItem {
        if (clipboardItem is LinkTreeItem) { // making link to link
            return clipboardItem.clone()
        }
        val link = LinkTreeItem(treeManager.currentItem, "", null)
        link.setTarget(treeClipboardManager.copiedFrom!!, clipboardItem.displayName)
        return link
    }

    fun copyAsText(text: String) {
        systemClipboardManager.copyToSystemClipboard(text)
        treeClipboardManager.clearClipboard()
        uiInfoService.showInfo("Text copied.")
    }
}